### PR TITLE
ddl: fix the covert job to rollingback job

### PR DESCRIPTION
**Pull request by AilinKid**
Signed-off-by: AilinKid <314806019@qq.com>

### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/23893

Once `convertJob2RollbackJob` meets error, the job state can't be set as `JobStateRollingback` since
job state and args may not be correctly overwritten. The job will be fetched to run with the canceling
state again. So we should check the error count here.

### What is changed and how it works?

What's Changed:
 
1. change all the rollingback assignments to the last step in all ddl canceling logic.   
**This is aim to get rid of invalid job state and job args, while the job is assigned with rollingback state, which will lead errors in following rollingback logic, for example, json unmarshal error.**

2. Add error count check and force it be canceled in convertJob2RollbackJob
**This is aim to get rid of the infinite loop of canceling action once `convertJob2RollbackJob` gets an error.**

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

- ddl: fix the covert job to rollingback job

### CODE REVIEW

File: ddl/rollingback.go [hunk 1]
```diff
@@ -447,12 +441,32 @@ func convertJob2RollbackJob(w *worker, d *ddlCtx, t *meta.Meta, job *model.Job)
...
+			//
+			// Once `convertJob2RollbackJob` meets error, the job state can't be set as `JobStateRollingback` since
+			// job state and args may not be correctly overwritten. The job will be fetched to run with the cancelling
+			// state again. So we should check the error count here.
+			if err1 := loadDDLVars(w); err1 != nil {
+				logutil.Logger(w.logCtx).Error("[ddl] load DDL global variable failed", zap.Error(err1))
+			}
+			errorCount := variable.GetDDLErrorCountLimit()
+			if job.ErrorCount > errorCount {
+				msg := fmt.Sprintf("job being converted to rollback job errors and error count beyond the limitation %d, cancelled", errorCount)
```
Comment (zimulala) [inline ddl/rollingback.go:463]:
```suggestion
				logutil.Logger(w.logCtx).Warn("[ddl] rollback DDL job error count exceed the limit, cancelled it now", zap.Int64("jobID", job.ID), zap.Int64("errorCountLimit", errorCount))
```

Unify the logs in `runDDL` and this function.

Comment (AilinKid) [reply ddl/rollingback.go]: addressed

File: ddl/rollingback.go [hunk 2]
```diff
@@ -447,12 +441,31 @@ func convertJob2RollbackJob(w *worker, d *ddlCtx, t *meta.Meta, job *model.Job)
...
 			job.Error = toTError(err)
 		}
-		if !job.Error.Equal(errCancelledDDLJob) {
-			job.Error = terror.GetErrClass(job.Error).Synthesize(terror.ErrCode(job.Error.Code()),
-				fmt.Sprintf("DDL job rollback, error msg: %s", terror.ToSQLError(job.Error).Message))
-		}
 		job.ErrorCount++
 
+		if errCancelledDDLJob.Equal(err) {
+			// job is normal cancelled.
```
Comment (tangenta) [inline ddl/rollingback.go:447]:
```suggestion
			// job is normally cancelled.
```

File: ddl/rollingback.go [hunk 3]
```diff
@@ -447,12 +441,31 @@ func convertJob2RollbackJob(w *worker, d *ddlCtx, t *meta.Meta, job *model.Job)
...
+			// Once `convertJob2RollbackJob` meets an error, the job state can't be set as `JobStateRollingback` since
+			// job state and args may not be correctly overwritten. The job will be fetched to run with the cancelling
+			// state again. So we should check the error count here.
+			if err1 := loadDDLVars(w); err1 != nil {
+				logutil.Logger(w.logCtx).Error("[ddl] load DDL global variable failed", zap.Error(err1))
+			}
+			errorCount := variable.GetDDLErrorCountLimit()
+			if job.ErrorCount > errorCount {
+				logutil.Logger(w.logCtx).Warn("[ddl] rollback DDL job error count exceed the limit, cancelled it now", zap.Int64("jobID", job.ID), zap.Int64("errorCountLimit", errorCount))
+				job.Error = toTError(errors.New(fmt.Sprintf(fmt.Sprintf(" rollback DDL job error count exceed the limit %d, cancelled it now", errorCount))))
```
Comment (tangenta) [inline ddl/rollingback.go:464]:
```suggestion
				job.Error = toTError(errors.New(fmt.Sprintf(fmt.Sprintf("rollback DDL job error count exceed the limit %d, cancelled it now", errorCount))))
```

**Review by tangenta — COMMENTED** [review]
rest LGTM

File: ddl/rollingback_test.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+	err = kv.RunInNewTxn(context.Background(), s.store, false, func(ctx context.Context, txn kv.Transaction) error {
+		t := meta.NewMeta(txn)
+		var err1 error
+		job, err1 = t.GetHistoryDDLJob(jobID)
+		return errors2.Trace(err1)
+	})
+	c.Assert(err, IsNil)
+	c.Assert(job.ErrorCount, Equals, int64(4))
+	c.Assert(job.Error.Error(), Equals, "[ddl:-1]rollback DDL job error count exceed the limit 3, cancelled it now")
+
```
Comment (zimulala) [inline ddl/rollingback_test.go:100]: Please remove this line.

File: ddl/rollingback.go [hunk 5]
```diff
@@ -447,12 +441,31 @@ func convertJob2RollbackJob(w *worker, d *ddlCtx, t *meta.Meta, job *model.Job)
...
 			job.Error = toTError(err)
 		}
-		if !job.Error.Equal(errCancelledDDLJob) {
-			job.Error = terror.GetErrClass(job.Error).Synthesize(terror.ErrCode(job.Error.Code()),
-				fmt.Sprintf("DDL job rollback, error msg: %s", terror.ToSQLError(job.Error).Message))
-		}
 		job.ErrorCount++
 
+		if errCancelledDDLJob.Equal(err) {
+			// job is normally cancelled.
```
Comment (zimulala) [inline ddl/rollingback.go:447]:
```suggestion
			// The job is normally cancelled.
```

File: ddl/rollingback.go [hunk 6]
```diff
@@ -447,12 +441,31 @@ func convertJob2RollbackJob(w *worker, d *ddlCtx, t *meta.Meta, job *model.Job)
...
+			// Once `convertJob2RollbackJob` meets an error, the job state can't be set as `JobStateRollingback` since
+			// job state and args may not be correctly overwritten. The job will be fetched to run with the cancelling
+			// state again. So we should check the error count here.
+			if err1 := loadDDLVars(w); err1 != nil {
+				logutil.Logger(w.logCtx).Error("[ddl] load DDL global variable failed", zap.Error(err1))
+			}
+			errorCount := variable.GetDDLErrorCountLimit()
+			if job.ErrorCount > errorCount {
+				logutil.Logger(w.logCtx).Warn("[ddl] rollback DDL job error count exceed the limit, cancelled it now", zap.Int64("jobID", job.ID), zap.Int64("errorCountLimit", errorCount))
+				job.Error = toTError(errors.New(fmt.Sprintf(fmt.Sprintf("rollback DDL job error count exceed the limit %d, cancelled it now", errorCount))))
```
Comment (zimulala) [inline ddl/rollingback.go:464]:
```suggestion
				job.Error = toTError(errors.Errorf("rollback DDL job error count exceed the limit %d, cancelled it now", errorCount))
```

File: ddl/rollingback_test.go [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+	tk := testkit.NewTestKit(c, s.store)
+	tk1 := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk1.MustExec("use test")
+
+	tk.MustExec("create table t_cancel_add_index (a int)")
+	tk.MustExec("insert into t_cancel_add_index values(1),(2),(3)")
+	tk.MustExec("set @@global.tidb_ddl_error_count_limit=3")
+
+	failpoint.Enable("github.com/pingcap/tidb/ddl/mockConvertAddIdxJob2RollbackJobError", `return(true)`)
```
Comment (zimulala) [inline ddl/rollingback_test.go:46]: We need to check the result of the function.

File: ddl/rollingback_test.go [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+	defer func() {
+		failpoint.Disable("github.com/pingcap/tidb/ddl/mockConvertAddIdxJob2RollbackJobError")
```
Comment (zimulala) [inline ddl/rollingback_test.go:48]: Ditto.

File: ddl/rollingback_test.go [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+		if job.Type != model.ActionAddIndex {
+			return
+		}
+		if job.SchemaState == model.StateDeleteOnly {
+			jobID = job.ID
+			res, checkErr = tk1.Exec("admin cancel ddl jobs " + strconv.Itoa(int(job.ID)))
+			// drain the result set here, otherwise the cancel action won't take effect immediately.
+			chk := res.NewChunk()
+			if err := res.Next(context.Background(), chk); err != nil {
+				checkErr = err
```
Comment (zimulala) [inline ddl/rollingback_test.go:74]: I think we need to return here.

**Pull request merged**